networkx/algorithms/coloring/tests/test_coloring.py

```diff
@@ -97,6 +102,202 @@ def test_seed_argument(self):
+            2: [0, 8],
+            3: [1, 7],
+            4: [1, 6],
+            5: [0, 6],
+            6: [4, 5],
+            7: [3],
+            8: [2],
+        }
+
+        F = {
+            # Color 0
+            0: 0,
+            1: 0,
+
+            # Color 1
+            2: 1,
+            3: 1,
+            4: 1,
+            5: 1,
+
+            # Color 2
+            6: 2,
+            7: 2,
+            8: 2,
+        }
+
+        C = nx.algorithms.coloring.equitable_coloring.make_C_from_F(F)
+        N = nx.algorithms.coloring.equitable_coloring.make_N_from_L_C(L, C)
+        H = nx.algorithms.coloring.equitable_coloring.make_H_from_C_N(C, N)
+
+        nx.algorithms.coloring.equitable_coloring.procedure_P(
+            V_minus=0, V_plus=1, N=N, H=H, F=F, C=C, L=L
+        )
+        check_state(L=L, N=N, H=H, F=F, C=C)
+
+    def test_cast_no_solo(self):
+        L = {
+            0: [8, 9],
+            1: [10, 11],
+
+            2: [8],
+            3: [9],
+            4: [10, 11],
+
+            5: [8],
+            6: [9],
+            7: [10, 11],
+
+            8: [0, 2, 5],
+            9: [0, 3, 6],
+            10: [1, 4, 7],
+            11: [1, 4, 7],
+        }
+
+        F = {
+            0: 0,
+            1: 0,
+
+            2: 2,
+            3: 2,
+            4: 2,
+
+            5: 3,
+            6: 3,
+            7: 3,
+
+            8: 1,
+            9: 1,
+            10: 1,
+            11: 1,
+        }
+
+        C = nx.algorithms.coloring.equitable_coloring.make_C_from_F(F)
+        N = nx.algorithms.coloring.equitable_coloring.make_N_from_L_C(L, C)
+        H = nx.algorithms.coloring.equitable_coloring.make_H_from_C_N(C, N)
+
+        nx.algorithms.coloring.equitable_coloring.procedure_P(
+            V_minus=0, V_plus=1, N=N, H=H, F=F, C=C, L=L
+        )
+        check_state(L=L, N=N, H=H, F=F, C=C)
+
+    def test_hard_prob(self):
+        # Tests for two levels of recursion.
+        num_colors, s = 5, 5
+
+        G = nx.Graph()
+        G.add_edges_from(
+            [(0, 10), (0, 11), (0, 12), (0, 23), (10, 4), (10, 9),
+             (10, 20), (11, 4), (11, 8), (11, 16), (12, 9), (12, 22),
+             (12, 23), (23, 7), (1, 17), (1, 18), (1, 19), (1, 24),
+             (17, 5), (17, 13), (17, 22), (18, 5), (19, 5), (19, 6),
+             (19, 8), (24, 7), (24, 16), (2, 4), (2, 13), (2, 14),
+             (2, 15), (4, 6), (13, 5), (13, 21), (14, 6), (14, 15),
+             (15, 6), (15, 21), (3, 16), (3, 20), (3, 21), (3, 22),
+             (16, 8), (20, 8), (21, 9), (22, 7)]
+        )
+        F = {node: node // s for node in range(num_colors * s)}
+        F[s - 1] = num_colors - 1
+
+        params = make_params_from_graph(G=G, F=F)
+
+        nx.algorithms.coloring.equitable_coloring.procedure_P(
+            V_minus=0, V_plus=num_colors - 1, **params
+        )
+        check_state(**params)
+
+    def test_hardest_prob(self):
+        # Tests for two levels of recursion.
```

Comment:
The graphs here were found using `Z3` which cause the algorithm to hit all different cases. The code to generate them is [here](https://github.com/musically-ut/equitable-coloring/blob/master/tests/generate_graph.py) and they are visualized [here](https://github.com/musically-ut/equitable-coloring/tree/master/tests/test-graphs).